### ISSUE TITLE
Feedback on EVA - Improvements to scripts and work flow listed in the document

### ISSUE THREAD

**Issue by JohnBMSFT** [MicrosoftDocs/azure-stack-docs#1950]
I have noticed few issues when working with a customer who's trying to test EVA feature in his environment. Suggesting the following corrections/updates to the document based on my observation:

**Issue#1:**
Use of "Read-Host" to pass the values to variables in "Operator enables a user subscription for EVA" section of the document is throwing below exception when trying to grant consent from tenant side

Invoke-AzResourceAction : OperationNotAllowed : The Emergency VM Access feature is not enabled for subscription '<subscription ID>'

**Suggestion:**
Although the actual cause for this is not known, assigning the values directly to variables isn't causing any issues. A work around here is to remove "Read-Host" in the commands and pass values directly like below:
$FQDN = "Enter External FQDN"
$RegionName = "Enter Azure Stack Region Name"
$TenantID = "Enter TenantID"
$TenantSubscriptionId  = "Enter Tenant Subscription ID"

Note: The cause for this being investigated and tracked via a work item created with engineering team

**Issue#2:**
The tenant ID to be provided for "$TenantID" variable in the operator section. This is not an issue with a single tenant environment as there is only a single tenant ID.
However, this causes confusion when customer is having a multi-tenancy environment as the document doesn't clearly capture the tenant ID to be provided. Providing tenant ID of any of the guest directory will throw following exception:

Login-AzAccount : InteractiveBrowserCredential authentication failed: AADSTS650057: Invalid resource. The client has requested access to a resource which is not listed in
the requested permissions in the client's application registration. Client app ID: <GUID>(Microsoft Azure PowerShell).

**Suggestion:**
The value for "$TenantID" should always be the tenant ID of home directory as it's only used for connecting to admin ARM endpoint. Adding a statement to provide the home directory tenant ID irrespective of the environment(single/multi-tenant) clears the confusion

**Issue#3:**
The ERCS VM to which PEP session has to be created. Executing the commands for connecting to the console in an ERCS VM other than the one listed by "$enableVMAccessResponse" command throws an exception

**Suggestion:**
Instead of having to enable RDP access for all ERCS VMs as suggested in the document(Note section of Operator enables remote desktop access to ERCS VMs), operator can create a PEP session only to the ERCS VM received from tenant (Output of "$enableVMAccessResponse")
The sections in the document can also be restructured for a clear workflow:

Work Flow:  Operator enables EVA feature for an user subscription -> Tenant user grants permission for the VM access -> Operator creates PEP session and launches the console

1. Operator enables a user subscription for EVA:

- No changes are needed

2. User to request VM console access

- Include only the points 1-2 which is to be performed by tenant user

3. Operator enables remote desktop access to ERCS VMs

- 	The section can be renamed appropriately and moved below tenant section as this activity needs to be performed based on the output from tenant's execution
- 	Must include the instructions for creating PEP and account to be used from existing section along with the points 3-9 from Tenant section
- 	Point 10 should contain the "Revoke-RdpAccessToErcsVM" command as this is the last step to be performed to harden the ERCS VM permissions

**Issue#4:**
There's an additional space in the script executed by tenant after "-ErrorAction Stop `" which throws an error

**Suggestion:**
Clearing the additional space from both Az and AzureRM script should resolve the issue

**Some additional feedback:**
Requesting the following changes to the script used by tenant to include:

1. The commands for connecting to ARM endpoint
2. Display the output that tenant user should share with operator:

- ERCS VM name
- Commands for Connecting to VM console which includes the resource ID 

**Note:** The $UserTenantID listed below should be the guest user's tenant ID. It'll be the same as home directory tenant ID in case of single tenant environment but will vary with each tenant in case of a multi-tenant environment

#Updated script for Azure RM modules

$RegionName = ""
$FQDN = ""
$UserTenantID = ""

#Add environment & authenticate
Add-AzureRmEnvironment -Name AzureStackUser -ARMEndpoint https://management.$RegionName.$FQDN
Login-AzureRmAccount -Environment AzureStackUser -TenantId $UserTenantID

$SubscriptionID= "your Azure subscription ID" 
$ResourceGroup = "your resource group name" 
$VMName = "your VM name" 
$vmResourceId = "/subscriptions/$SubscriptionID/resourceGroups/$ResourceGroup/providers/Microsoft.Compute/virtualMachines/$VMName" 

$enableVMAccessResponse = Invoke-AzureRMResourceAction `
    -ResourceId $vmResourceId `
    -Action "enableVmAccess" `
    -ApiVersion "2020-06-01" `
    -ErrorAction Stop `
    -Force 

Write-Host "Please provide the following output to operator`n" -ForegroundColor Yellow

Write-Host "ERCS Name:`t$(($enableVMAccessResponse).ERCSName)" -ForegroundColor Cyan
Write-Host "ConnectTo-TenantVm -ResourceID $($vmResourceId)" -ForegroundColor Green
Write-Host "Delete-TenantVMSession -ResourceID $($vmResourceId)" -ForegroundColor Green

#################################################################################################################

#Updated Script for AZ modules

$RegionName = ""
$FQDN = ""
$UserTenantID = ""

#Add environment & authenticate
Add-AzEnvironment -Name AzureStackUser -ARMEndpoint https://management.$RegionName.$FQDN
Login-AzAccount -Environment AzureStackUser -TenantId $UserTenantID

$SubscriptionID= "your Azure subscription ID" 
$ResourceGroup = "your resource group name" 
$VMName = "your VM name" 
$vmResourceId = "/subscriptions/$SubscriptionID/resourceGroups/$ResourceGroup/providers/Microsoft.Compute/virtualMachines/$VMName" 

$enableVMAccessResponse = Invoke-AzResourceAction `
    -ResourceId $vmResourceId `
    -Action "enableVmAccess" `
    -ApiVersion "2020-06-01" `
    -ErrorAction Stop `
    -Force 

Write-Host "Please provide the following output to operator`n" -ForegroundColor Yellow

Write-Host "ERCS Name:`t$(($enableVMAccessResponse).ERCSName)" -ForegroundColor Cyan
Write-Host "ConnectTo-TenantVm -ResourceID $($vmResourceId)" -ForegroundColor Green
Write-Host "Delete-TenantVMSession -ResourceID $($vmResourceId)" -ForegroundColor Green
#################################################################################################################

Lastly, the EVA feature is limited to Windows servers running with a GUI version as core operating system doesn't support On-screen keyboard functionality. So it's impossible for a customer to login to a core server even though he can connect to it's console as Ctrl+Alt+Del key combination cannot be sent as an input. Customer needs to engage Microsoft support to provide console access from unlock PEP in the event he needs to address an issue with Windows Core OS. A note can be added in the document that outlines this statement.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 221ace5c-e2df-f784-d570-a39563c13a8d
* Version Independent ID: 38a14acb-21f5-5378-6ece-50e89e437025
* Content: [Emergency VM access in Azure Stack Hub - Azure Stack Hub](https://docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-2108&tabs=azurerm1)
* Content Source: [azure-stack/user/emergency-vm-access.md](https://github.com/MicrosoftDocs/azure-stack-docs/blob/main/azure-stack/user/emergency-vm-access.md)
* Service: **azure-stack**
* Sub-service: **azure-stack-hub**
* GitHub Login: @sethmanheim
* Microsoft Alias: **sethm**

**Comment by VikasPullagura-MSFT** [comment]
@sethmanheim 
Thanks for the feedback! I have assigned the issue to content author to check and update the document as appropriate.

**Comment by VikasPullagura-MSFT** [comment]
@sethmanheim 
Can you please check and add your comments on this doc update request as applicable.

**Comment by sethmanheim** [comment]
@JohnBMSFT I have made some of the fixes suggested here but will need more information about what you're suggesting in the "Some additional feedback" section. There isn't enough information here to make updates (i.e. what sections and scripts you're referring to).

**Comment by JohnBMSFT** [comment]
Hi Seth,

Providing below some additional information on the suggestions I requested for the EVA document. Please let me know if you need any additional information.

Issue#4:
There's an additional space in the script executed by tenant after "-ErrorAction Stop `" and throws an error. Although it is not preventing the script from execution it just throws below error.

Error Details: Space highlighted in line 10 is the cause for below error
***@***.***

Suggestion:
Clearing the additional space from both Az and AzureRM scripts under "Tenant<https://docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-2108&tabs=azurerm1#user-to-request-vm-console-access>" section should resolve the issue
###################################################################################################################################################################

Some additional feedback:
Requesting the following changes to the script used by tenant to include:

#1: The commands for connecting to ARM endpoint
Current Scenario:
The existing document<https://docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-2108&tabs=azurerm1#user-to-request-vm-console-access> doesn't contain any commands to connect to tenant ARM endpoint. It simply suggests to open PowerShell and sign-in to the subscription which might be confusing to any new customer.
***@***.***

Suggestion:
We can either include a link pointing to the below document in the first point asking customer to connect to tenant ARM endpoint using PowerShell or include the commands from the link directly in the script.
Connect to Azure Stack Hub with PowerShell as a user - Azure Stack Hub | Microsoft Docs<https://docs.microsoft.com/en-us/azure-stack/user/azure-stack-powershell-configure-user?view=azs-2108&tabs=az1%2Caz2%2Caz3%2Caz4>

The commands I included in the GitHub request contains a variant of the commands listed in the above document. I'd prefer leaving the existing script intact and adding the above link to point#1.

###################################################################################################################################################################
#2: Display the output that tenant user should share with operator

  *   ERCS VM name
  *   Commands for Connecting to VM console which includes the resource ID

Current Scenario:
The script discussed above doesn't return any output. However, the point 2 from document suggests tenant user to provide the details of ERCS VM Name and VMResourceID to operator.

***@***.***

***@***.***

Suggestion:
I've suggested to include the following lines at the end of the script so tenant user can share the output to operator which contains the commands needed by operator at a later stage(Snip from doc) along with the ERCS VM name.

Write-Host "Please provide the following output to operator`n" -ForegroundColor Yellow
Write-Host "ERCS Name:`t$(($enableVMAccessResponse).ERCSName)" -ForegroundColor Cyan
Write-Host "ConnectTo-TenantVm -ResourceID $($vmResourceId)" -ForegroundColor Green
Write-Host "Delete-TenantVMSession -ResourceID $($vmResourceId)" -ForegroundColor Green

***@***.***

Script Output with above lines included:
***@***.***

###################################################################################################################################################################
#3: EVA limitation with Windows Core OS:
Lastly, the EVA feature is limited to Windows servers running with a GUI version as core operating system doesn't support On-screen keyboard functionality. So it's impossible for a customer to login to a core server even though he can connect to it's console as Ctrl+Alt+Del key combination cannot be sent as an input. Customer needs to engage Microsoft support to provide console access from unlock PEP in the event he needs to address an issue with Windows Core OS.

We cannot pass Windows + U key combination here to launch on-screen keyboard as Core OS doesn't support it.
***@***.***

Suggestion:
A note can be added in the article that outlines the above limitation with Windows Core OS.

###################################################################################################################################################################
I had gone through the changes made so far to the document using the below link:

https://review.docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-1802&branch=pr-en-us-10265&tabs=azurerm1

The tenant section<https://review.docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-1802&branch=pr-en-us-10265&tabs=azurerm1#user-to-request-vm-console-access> just needs changes to the script and point#1 which I suggested above.
The title for operator section<https://review.docs.microsoft.com/en-us/azure-stack/user/emergency-vm-access?view=azs-1802&branch=pr-en-us-10265&tabs=azurerm1#operator-enables-remote-desktop-access-to-ercs-vms> can be renamed appropriately to cover the actions being performed(For example: Operator enables remote desktop access to ERCS VMs and launches VM console) and first paragraph can be updated as shown below.
Point#1 can include a statement to connect only to the ERCS VM provided by tenant user and the "Note" section can be discarded.
Also the reason for suggesting to move the "Revoke-RdpAccessToErcsVM" command to the end is because this should be the last step to be performed to harden the ERCS VM permissions once end user is done with his work through the VM console.
***@***.***

###################################################################################################################################################################

Best Regards,
John B.

Support Engineer
Microsoft Azure Stack Hub Technical Support
Customer Service and Support

***@***.******@***.***>
Working hours: Mon-Fri 13:30 - 22:30 IST (GMT+5:30)
If you are looking for support outside of my working hours, please send an email to Azure Stack Hub CSS ***@***.***> with your case number and availability.
We are always interested to hear your feedback. Please feel free to reach my manager regarding the level of service you have received. ***@***.******@***.***>
[MSFT_logo]

**Comment by sethmanheim** [comment]
@JohnBMSFT Thank you for taking the time to write this up. It seems clear, and I will work on these changes ASAP.

**Comment by sethmanheim** [comment]
All the changes have been published, but feel free to open a new issue, if anything else pops up. Contacting me internally would be preferable, FYI. :-) #please-close
